File: commands/src/main/java/kaap/veiko/debuggerforker/commands/commandsets/event/events/ClassPrepareEvent.java
package kaap.veiko.debuggerforker.commands.commandsets.event.events;

import com.google.auto.value.AutoValue;

import kaap.veiko.debuggerforker.types.DataReadException;
import kaap.veiko.debuggerforker.types.DataReader;
import kaap.veiko.debuggerforker.types.DataWriter;
import kaap.veiko.debuggerforker.types.jdwp.EventKind;
import kaap.veiko.debuggerforker.types.jdwp.ReferenceTypeId;
import kaap.veiko.debuggerforker.types.jdwp.ThreadId;

@AutoValue
public abstract class ClassPrepareEvent extends VirtualMachineEvent {
  public abstract int getRequestId();
  public abstract ThreadId getThread();
  public abstract byte getRefTypeTag();
  public abstract ReferenceTypeId getTypeId();
  public abstract String getSignature();
  public abstract int getStatus();

  public EventKind getEventKind() {
    return EventKind.CLASS_PREPARE;
  }

  public static ClassPrepareEvent create(int requestId, ThreadId thread, byte refTypeTag, ReferenceTypeId typeId, String signature, int status) {
    return new AutoValue_ClassPrepareEvent(requestId, thread, refTypeTag, typeId, signature, status);
  }

  public static ClassPrepareEvent read(DataReader reader) throws DataReadException {
    return create(
        reader.readInt(),
        ThreadId.read(reader),
        reader.readByte(),
        ReferenceTypeId.read(reader),
        reader.readString(),
        reader.readInt()
    );
  }

  @Override
  public void write(DataWriter writer) {
    writer.writeType(getEventKind());
    writer.writeInt(getRequestId());
    writer.writeType(getThread());
    writer.writeByte(getRefTypeTag());
    writer.writeType(getTypeId());
    writer.writeString(getSignature());
    writer.writeInt(getStatus());
  }
}

File: commands/src/main/kotlin/kaap/veiko/debuggerforker/commands/commandsets/event/events/ClassPrepareEvent.kt
package kaap.veiko.debuggerforker.commands.commandsets.event.events

import kaap.veiko.debuggerforker.types.DataReader
import kaap.veiko.debuggerforker.types.DataWriter
import kaap.veiko.debuggerforker.types.jdwp.EventKind
import kaap.veiko.debuggerforker.types.jdwp.ReferenceTypeId
import kaap.veiko.debuggerforker.types.jdwp.ThreadId

data class ClassPrepareEvent(
        val requestId: Int,
        val thread: ThreadId,
        val refTypeTag: Byte,
        val typeId: ReferenceTypeId,
        val signature: String,
        val status: Int
) : VirtualMachineEvent() {

    override fun write(writer: DataWriter?) = writer!!.run {
        writeType(EventKind.CLASS_PREPARE)
        writeInt(requestId)
        writeType(thread)
        writeByte(refTypeTag)
        writeType(typeId)
        writeString(signature)
        writeInt(status)
    }

    companion object {
        fun read(reader: DataReader) = ClassPrepareEvent(
                reader.readInt(),
                ThreadId.read(reader),
                reader.readByte(),
                ReferenceTypeId.read(reader),
                reader.readString(),
                reader.readInt()
        )
    }
}